src/dense_vector.rs

```diff
@@ -0,0 +1,125 @@
+use crate::Ix1;
+use ndarray::{self, ArrayBase};
+
+/// A trait for types representing dense vectors, useful for
+/// defining a fast sparse-dense dot product.
+pub trait DenseVector<N> {
+    /// The dimension of the vector
+    fn dim(&self) -> usize;
+
+    /// Random access to an element in the vector.
+    ///
+    /// # Panics
+    ///
+    /// If the index is out of bounds
+    fn index(&self, idx: usize) -> &N;
+}
+
+impl<'a, N: 'a> DenseVector<N> for &'a [N] {
+    fn dim(&self) -> usize {
+        self.len()
+    }
+
+    #[inline]
```

Comment:
I think we should use `#[inline(always)]` on these, they should be zero-cost

Reply:
You're right.